File: app/candles/timeframe_candle_engine.py
from collections import defaultdict
from datetime import datetime

import orjson

from app.cache.redis import redis_client
from app.core.logger import logger
from app.services.candle.candle_publisher import CandlePublisher


class TimeframeCandleEngine:

    candles = defaultdict(dict)

    TIMEFRAMES = [
        "1m",
        "5m",
        "15m"
    ]

    @classmethod
    def _is_complete(
        cls,
        candle: dict
    ):

        age = (
            datetime.utcnow()
            - candle["created_at"]
        ).seconds

        timeframe = candle["timeframe"]

        if timeframe == "1m":
            return age >= 60

        if timeframe == "5m":
            return age >= 300

        if timeframe == "15m":
            return age >= 900

        return False

    @classmethod
    async def _publish_live(cls, candle: dict):
        """
        Store the current in-progress candle in Redis so the FastAPI
        WebSocket endpoint (different process) can read and stream it.
          Key:  live_candle:{symbol}:{timeframe}  (TTL 120s)
          PubSub channel: live_candle_events:{symbol}:{timeframe}
        """
        redis_key = f"live_candle:{candle['symbol']}:{candle['timeframe']}"
        channel = f"live_candle_events:{candle['symbol']}:{candle['timeframe']}"

        payload = orjson.dumps({
            "symbol": candle["symbol"],
            "timeframe": candle["timeframe"],
            "open": candle["open"],
            "high": candle["high"],
            "low": candle["low"],
            "close": candle["close"],
            "volume": candle["volume"],
            "timestamp": candle["created_at"].isoformat(),
            "is_live": True,
        }).decode()

        # Store latest state (readable by HTTP endpoint)
        await redis_client.set(redis_key, payload, ex=120)

        # Pub/sub push so WebSocket can forward to browser immediately
        await redis_client.publish(channel, payload)

    @classmethod
    async def process_tick(
        cls,
        symbol,
        price,
        quantity=0
    ):

        for timeframe in cls.TIMEFRAMES:

            key = f"{symbol}:{timeframe}"

            candle = cls.candles.get(key)

            if candle and cls._is_complete(candle):

                completed_candle = {
                    "symbol": candle["symbol"],
                    "timeframe": candle["timeframe"],
                    "open": candle["open"],
                    "high": candle["high"],
                    "low": candle["low"],
                    "close": candle["close"],
                    "volume": candle["volume"],
                    "timestamp": candle["created_at"].isoformat()
                }

                await CandlePublisher.publish(completed_candle)

                logger.info(f"Published Candle: {key}")

                # Start a new candle from this tick
                cls.candles[key] = {
                    "symbol": symbol,
                    "timeframe": timeframe,
                    "open": price,
                    "high": price,
                    "low": price,
                    "close": price,
                    "volume": quantity,
                    "created_at": datetime.utcnow()
                }

                await cls._publish_live(cls.candles[key])
                continue

            if not candle:

                cls.candles[key] = {
                    "symbol": symbol,
                    "timeframe": timeframe,
                    "open": price,
                    "high": price,
                    "low": price,
                    "close": price,
                    "volume": quantity,
                    "created_at": datetime.utcnow()
                }

            else:

                candle["high"] = max(candle["high"], price)
                candle["low"] = min(candle["low"], price)
                candle["close"] = price
                candle["volume"] += quantity

            # Publish live state on every tick update
            await cls._publish_live(cls.candles[key])

            logger.info(f"{key} Candle Updated")

File: backend/src/trading/candles/timeframeCandleEngine.ts
import redisClient from "../../core/redis";
import { logger } from "../../core/logger";
import { CandlePublisher } from "../../services/candlePublisher";

interface CandleState {
  symbol: string;
  timeframe: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  created_at: Date;
}

export class TimeframeCandleEngine {
  private static candles: { [key: string]: CandleState } = {};
  public static TIMEFRAMES = ["1m", "5m", "15m"];

  private static isComplete(candle: CandleState): boolean {
    const ageSeconds = Math.floor((Date.now() - candle.created_at.getTime()) / 1000);
    const timeframe = candle.timeframe;

    if (timeframe === "1m") return ageSeconds >= 60;
    if (timeframe === "5m") return ageSeconds >= 300;
    if (timeframe === "15m") return ageSeconds >= 900;
    return false;
  }

  private static async publishLive(candle: CandleState) {
    const redisKey = `live_candle:${candle.symbol}:${candle.timeframe}`;
    const channel = `live_candle_events:${candle.symbol}:${candle.timeframe}`;

    const payload = JSON.stringify({
      symbol: candle.symbol,
      timeframe: candle.timeframe,
      open: candle.open,
      high: candle.high,
      low: candle.low,
      close: candle.close,
      volume: candle.volume,
      timestamp: candle.created_at.toISOString(),
      is_live: true,
    });

    // Store latest state (readable by HTTP endpoint)
    await redisClient.set(redisKey, payload, "EX", 120);

    // Pub/sub push so WebSocket can forward to browser immediately
    await redisClient.publish(channel, payload);
  }

  static async processTick(symbol: string, price: number, quantity: number = 0) {
    for (const timeframe of this.TIMEFRAMES) {
      const key = `${symbol}:${timeframe}`;
      const candle = this.candles[key];

      if (candle && this.isComplete(candle)) {
        const completedCandle = {
          symbol: candle.symbol,
          timeframe: candle.timeframe,
          open: candle.open,
          high: candle.high,
          low: candle.low,
          close: candle.close,
          volume: candle.volume,
          timestamp: candle.created_at.toISOString(),
        };

        await CandlePublisher.publish(completedCandle);
        logger.info(`Published Candle: ${key}`);

        // Start a new candle from this tick
        this.candles[key] = {
          symbol,
          timeframe,
          open: price,
          high: price,
          low: price,
          close: price,
          volume: quantity,
          created_at: new Date(),
        };

        await this.publishLive(this.candles[key]);
        continue;
      }

      if (!candle) {
        this.candles[key] = {
          symbol,
          timeframe,
          open: price,
          high: price,
          low: price,
          close: price,
          volume: quantity,
          created_at: new Date(),
        };
      } else {
        candle.high = Math.max(candle.high, price);
        candle.low = Math.min(candle.low, price);
        candle.close = price;
        candle.volume += quantity;
      }

      // Publish live state on every tick update
      await this.publishLive(this.candles[key]);
      logger.info(`${key} Candle Updated`);
    }
  }
}
export default TimeframeCandleEngine;
